src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorLSPTextViewConnectionListener.cs

```diff
@@ -63,11 +121,131 @@ public void SubjectBuffersConnected(ITextView textView, ConnectionReason reason,
+                    textView.TextBuffer.IsRazorLSPBuffer())
+                {
+                    // We assume there is ever only one TextBuffer at a time and thus all active
+                    // TextViews have the same TextBuffer.
+                    _textBuffer = textView.TextBuffer;
+
+                    var bufferOptions = _editorOptionsFactory.GetOptions(_textBuffer);
+                    _textBuffer.Properties[RazorEditorBufferOptions] = bufferOptions;
+
+                    // All TextViews share the same options, so we only need to listen to changes for one.
+                    _textBuffer.Properties[RazorEditorTrackedView] = textView;
+
+                    var razorEditorTextViewOptions = _editorOptionsFactory.GetOptions(textView);
+                    Assumes.Present(razorEditorTextViewOptions);
+                    _textBuffer.Properties[RazorEditorViewOptions] = razorEditorTextViewOptions;
+
+                    RazorOptions_OptionChanged(null, null);
+                    razorEditorTextViewOptions.OptionChanged += RazorOptions_OptionChanged;
+                }
+            }
         }
 
         public void SubjectBuffersDisconnected(ITextView textView, ConnectionReason reason, IReadOnlyCollection<ITextBuffer> subjectBuffers)
         {
             // When the TextView goes away so does the filter.  No need to do anything more.
+            // However, we do need to detach from listening for option changes to avoid leaking.
+            // We should switch to listening to a different TextView if the one we're listening
+            // to is disconnected.
+            Assumes.NotNull(_textBuffer);
+
+            lock (_lock)
+            {
+                _activeTextViews.Remove(textView);
+
+                // Is the tracked TextView where we listen for option changes the one being disconnected?
+                // If so, see if another view is available.
+                if (_textBuffer.Properties.TryGetProperty(RazorEditorTrackedView, out ITextView trackedTextView) &&
+                    trackedTextView == textView)
+                {
+                    Assumes.True(_textBuffer.Properties.TryGetProperty(RazorEditorViewOptions, out IEditorOptions textViewOptions));
+
+                    // If there's another text view we can use to listen for options, start tracking it.
+                    _textBuffer.Properties.RemoveProperty(RazorEditorTrackedView);
+                    _textBuffer.Properties.RemoveProperty(RazorEditorViewOptions);
+                    textViewOptions.OptionChanged -= RazorOptions_OptionChanged;
+
+                    if (_activeTextViews.Count != 0)
+                    {
+                        var updatedTextView = _activeTextViews[0];
+                        _textBuffer.Properties[RazorEditorTrackedView] = updatedTextView;
+
+                        var updatedRazorEditorTextViewOptions = _editorOptionsFactory.GetOptions(updatedTextView);
```

Comment:
Very interesting, so it definitely seems like the TextView option tracking is important. Will be curious to understand what this piece does. I'm sure this was a fun one to optimize against 😆